File: program/program/find-average-of-numbers-by-recursion/FindAverageOfNumbersByRecursion.swift
import Foundation
/*

This Swift program calculates the average of a list of numbers using recursion. The user is prompted to input a series of numbers separated by spaces. The program then employs a recursive algorithm to compute the sum of the provided numbers, and subsequently, calculates and displays their average. In the event of invalid input, the program notifies the user and prompts for a valid list of numbers. This implementation is designed for simplicity and educational purposes, offering a basic example of recursive function usage in Swift.

Sample Input:
    Enter a list of numbers separated by spaces:
    5.0 10.0 15.0 20.0

Sample Output:
    Numbers: [5.0, 10.0, 15.0, 20.0]
    Average: 12.5

*/



func calculateAverage(numbers: [Double]) -> Double {
    return sumOfNumbers(numbers: numbers) / Double(numbers.count)
}

func sumOfNumbers(numbers: [Double]) -> Double {
    guard let firstNumber = numbers.first else {
        return 0
    }
    let remainingNumbers = Array(numbers.dropFirst())
    return firstNumber + sumOfNumbers(numbers: remainingNumbers)
}

// Get input from the user
print("Enter a list of numbers separated by spaces:")
if let input = readLine(), !input.isEmpty {
    let numbers = input.split(separator: " ").compactMap { Double($0) }
    
    if !numbers.isEmpty {
        let average = calculateAverage(numbers: numbers)
        print("Numbers: \(numbers)")
        print("Average: \(average)")
    } else {
        print("Invalid input. Please enter valid numbers.")
    }
} else {
    print("Invalid input. Please enter valid numbers.")
}

File: program/program/find-average-of-numbers-by-recursion/FindAverageOfNumbersByRecursion.rb
def calculate_average(numbers, index = 0)
  if index == numbers.length - 1
    return numbers[index].to_f
  else
    return numbers[index].to_f + calculate_average(numbers, index + 1)
  end
end

def main
  puts "Enter numbers separated by spaces:"
  input = gets.chomp
  numbers = input.split.map(&:to_i)

  if numbers.empty?
    puts "No numbers entered. Exiting program."
  else
    average = calculate_average(numbers) / numbers.length
    puts "The average of #{numbers.join(', ')} is: #{average}"
  end
end

main
